rust/noosphere-core/src/data/body_chunk.rs

```diff
@@ -70,4 +76,326 @@ impl BodyChunkIpld {
+        self.data.push(chunk).await?;
+        Ok(())
+    }
+
+    /// Consumes this [BodyChunkEncoder], encoding into the store, returning
+    /// the root [Cid] upon success.
+    pub async fn encode(mut self) -> Result<Cid> {
+        let mut next_chunk_cid = None;
+        let chunk_stream = self.data.into_stream_rev().await;
+        tokio::pin!(chunk_stream);
+
+        while let Some(chunk) = chunk_stream.try_next().await? {
+            next_chunk_cid = Some(
+                self.store
+                    .save::<DagCborCodec, _>(&BodyChunkIpld {
+                        bytes: chunk,
+                        next: next_chunk_cid,
+                    })
+                    .await?,
+            );
+        }
+        Ok(next_chunk_cid.unwrap())
+    }
+}
+
+/// Container for short-term storage of sequenced blobs
+/// for consumption as a forward or reverse stream.
+/// [BlobVec] data is backed by memory, dynamically converting
+/// to disk storage if memory threshold is reached (on non-`wasm32` platforms).
+/// Blobs can then be drained via `BlobVec::into_stream`.
+struct BlobVec {
+    chunks: BlobVecData,
+    memory_limit: u64,
+}
+
+impl BlobVec {
+    /// Creates a new [BlobVec] that stores data in memory
+    /// until `memory_limit` is reached.
+    pub fn new(memory_limit: u64) -> Self {
+        BlobVec {
+            chunks: BlobVecData::new_with_memory(),
+            memory_limit,
+        }
+    }
+
+    /// Push a new item.
+    pub async fn push(&mut self, chunk: Vec<u8>) -> Result<()> {
+        self.chunks.push(chunk).await?;
+
+        #[cfg(not(target_arch = "wasm32"))]
+        if matches!(self.chunks, BlobVecData::Memory { .. })
+            && self.chunks.byte_length() > &self.memory_limit
+        {
+            self.upgrade_storage().await?;
+        }
+
+        Ok(())
+    }
+
+    /// Drain all items into a stream.
+    #[allow(unused)]
+    pub async fn into_stream(self) -> impl Stream<Item = Result<Vec<u8>, std::io::Error>> + Unpin {
+        self.chunks.into_stream(false).await
+    }
+
+    /// Drain all items into a stream in reverse.
+    pub async fn into_stream_rev(
+        self,
+    ) -> impl Stream<Item = Result<Vec<u8>, std::io::Error>> + Unpin {
+        self.chunks.into_stream(true).await
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn upgrade_storage(&mut self) -> Result<()> {
+        let mut storage = BlobVecData::new_with_disk()?;
+        std::mem::swap(&mut self.chunks, &mut storage);
+        let previous_stream = storage.into_stream(false).await;
+        tokio::pin!(previous_stream);
+        while let Some(chunk) = previous_stream.try_next().await? {
+            self.chunks.push(chunk).await?;
+        }
+        Ok(())
+    }
+}
+
+enum BlobVecData {
+    Memory {
+        chunks: Vec<Vec<u8>>,
+        chunk_count: usize,
+        byte_length: u64,
+    },
+    #[cfg(not(target_arch = "wasm32"))]
+    Disk {
```

Comment:
If we used a `Store`-based construct as our fundamental primitive for this, we could be generic to both web and native. On web, the scratch space would be a object store in IndexedDB, and on native it would be a Sled tree.

Reply:
Just observing that encoding a large file will be no more viable on the web than it is on the desktop (and in fact we are probably _even more_ sensitive to our memory ceiling on the web).

Reply:
ya SGTM!